### ISSUE TITLE
HM.Files.ClassFileName.MismatchedName includes interfaces

### ISSUE THREAD

**Issue by dan-westall** [humanmade/coding-standards#11]
This sniff currently includes the token for interfaces, and errors, insisting that the file have a class prefix on the file name. I dont think this is correct behaviour a interface filename should be `interface-foo.php` not `class-foo.php`

Is this intentional or a bug, if its intentional whats the reasoning behind it.

**Comment by rmccue** [comment]
This is due to the way autoloading works in PHP. When your autoloader gets called, it receives the class name, and nothing else. For the purposes of autoloading, interfaces and traits are also handled the exact same way.

Because of that, there's no way in the autoloader to differentiate between class/interface/trait, so you can't change the naming up based on the type. If we wanted to have `interface-X.php` work in the autoloader, we'd need to check for two files for every class: `class-X.php` and `interface-X.php`. Giving traits a separate name would add another one to the list. Since the autoloader gets called a *lot*, the performance matters, and hitting the stat cache this much would be a bit too slow.

That said, you can design autoloaders to allow this: class map autoloaders generated by a build tool would work fine with this (and have better performance generally), but they introduce an extra build step.

tl;dr Since the only autoloader we have is a dynamic one (receives a name and *generates* the filename at runtime), we can't distinguish between classes, interfaces, and traits without affecting performance.

**Comment by rmccue** [comment]
Closing this out, as I don't think we need to change it based on the above. :)
